### ISSUE TITLE
ClickHouse to Kafka: add the possibility to specify a partition key for Kafka messages

### ISSUE THREAD

**Issue by victor-isaev** [ClickHouse/ClickHouse#55810]
**Use case**

I have a `test` table in ClickHouse that I want to populate to Kakfa with Kafka Engine. I need any changes to `test` to be pushed to Kafka, so I've configured a materialized view for that:

```sql
CREATE TABLE test(key String, value INT)
ENGINE = MergeTree()
ORDER BY (key, value);

CREATE MATERIALIZED VIEW test_mv
ENGINE = Kafka('broker:9092', 'test', 'test_group', 'JSONEachRow')
AS SELECT * FROM test;

INSERT INTO test VALUES ('X', 1);
```

This works great, and the messages arrive to my Kafka instance. However, the messages do not have keys:

<img width="610" alt="image" src="https://github.com/ClickHouse/ClickHouse/assets/138467999/909362ff-fd0e-4a50-a861-10297f08b756">

**Describe the solution you'd like**

I would like to be able to specify the partitioning key for the messages that are sent by ClickHouse to Kafka. This way I will be able to distribute the load between multiple consumers of those messages while still having the quarantee that the messages with the same key arrive to the same consumer strictly in order they happened.

Syntactically, we could allow the `PARTITION BY` clause in Kafka Engine, like this:

```sql
CREATE MATERIALIZED VIEW test_mv
ENGINE = Kafka('broker:9092', 'test', 'test_group', 'JSONEachRow')
PARTITION BY key
AS SELECT * FROM test;
```

This way we could specify that the values from the `key` column in our `test` table should be used as a partition key for outbound messages.

**Describe alternatives you've considered**

The current way to handle this is to collect all the outbound messages in the topic with a single partition (this guarantees they are not reordered), and then repartition it to another topic by using either a hand-made consumer or a thing like [ksqlDB](https://ksqldb.io/) or [Flink](https://flink.apache.org/). In the mentioned solution our first topic is still a bottleneck, because it has only 1 partition, so we can only handle it with a single consumer - no scalability.

**Comment by sigxcpu76** [comment]
https://github.com/ClickHouse/ClickHouse/issues/2691

**Comment by jovezhong** [comment]
As https://github.com/ClickHouse/ClickHouse/issues/2691#issuecomment-1266855303, this is doable by declaring the _key in the DDL and INSERT to the column. However I guess you just need a key for the Kafka message, not necessary the partition. Maybe this ticket can be closed.

**Comment by ahmadkkhaled** [comment]
@jovezhong I don't agree that this ticket can be closed. The author is interested in sharding/distributing the clickhouse kafka consumer work across multiple clickhouse instances.

For example, creating a table with a kafka engine that reads strictly from a specific set of partitions belonging to a single topic.

To my understanding, the solution of writing to the _key does not address the scalability issue

**Comment by jovezhong** [comment]
Ideally we can allow the developer to specify the kafka message key and/or partition id. Here is how things work in Timeplus (a streaming engine built on top of ClickHouse)

```sql
CREATE EXTERNAL STREAM foo (
    id int32,
    name string,
    _tp_message_key nullable(string) default null
) SETTINGS type='kafka',...;

INSERT INTO foo(id,name,_tp_message_key) VALUES (1, 'John', 'some-key');

CREATE EXTERNAL STREAM foo (
    id int32,..
) SETTINGS type='kafka', sharding_expr='hash(id)'...;
--you can hardcode the value in sharding_expr if you don't want to use a mapping function

```

More:  https://docs.timeplus.com/proton-kafka#metadata

**Comment by alexey-milovidov** [comment]
This is now possible without a dedicated `PARTITION BY` clause.

When producing to Kafka, the engine maps a `String` column named `_key` to the Kafka message **key**, and Kafka's default partitioner routes a message by `murmur2(key) % num_partitions`. So setting `_key` to your partitioning expression gives exactly the behavior requested here: messages with the same key always land on the same partition, preserving per-key order while still letting you scale consumers across partitions.

The recent setting `kafka_map_virtual_columns_on_write` (added in #103243) makes this clean by excluding the `_key`/`_timestamp`/`_headers` columns from the produced message payload, so the key no longer leaks into the JSON body.

Your example becomes:

```sql
CREATE TABLE test (key String, value Int32)
ENGINE = MergeTree ORDER BY (key, value);

CREATE TABLE test_kafka (`_key` String, key String, value Int32)
ENGINE = Kafka('broker:9092', 'test', 'test_group', 'JSONEachRow')
SETTINGS kafka_map_virtual_columns_on_write = 1;

CREATE MATERIALIZED VIEW test_mv TO test_kafka
AS SELECT key AS _key, key, value FROM test;

INSERT INTO test VALUES ('X', 1);
```

Here `key AS _key` plays the role of the proposed `PARTITION BY key`: the value of `key` becomes the message key, the partitioner derives the partition from it, and with `kafka_map_virtual_columns_on_write = 1` the payload stays `{"key":"X","value":1}` without an extra `_key` field.

The same mechanism also lets you set the message `_timestamp` and `_headers` on write — see the [Mapping columns to Kafka message metadata](https://clickhouse.com/docs/engines/table-engines/integrations/kafka) docs section.

Closing as resolved. (Note: choosing an explicit partition *number* directly — rather than by key — is a separate, smaller enhancement; if anyone needs that, please open a new issue.)
